src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -765,8 +772,17 @@ void java_lang_String::print(oop java_string, outputStream* st) {
   int length = java_lang_String::length(java_string, value);
   bool is_latin1 = java_lang_String::is_latin1(java_string);
 
+  bool abridge = length > max_length;
+
   st->print("\"");
   for (int index = 0; index < length; index++) {
+    // If we need to abridge and we've printed half the allowed characters
+    // then jump to the tail of the string.
+    if (abridge && index >= max_length / 2) {
+      st->print(" ... ");
+      index = length - (max_length / 2);
+      abridge = false; // only do this once
+    }
```

Comment:
Instead of the trailing "abridged", in similar cases I printed out the number of omitted characters. E.g.

"Very long long long ... (53 characters omitted) ... long long string"

Makes it obvious how much has been cut, and no danger of confusing the ellipse with naturally occurring dots.

Additionally, I would only do this if length > max_length + X, with X being at least as long as the middle part (3 characters if you only print an ellipse). You end up with printed strings that may be slightly longer than maxlen, but OTOH the output is clearer. Otherwise you may indicate omission where none happened (if length == max_length)

Reply:
@tstuefe  - thanks for looking at this Thomas. I don't get your second point. First I only abridge when length > max_length. Second adding in the X fudge factor just means max_length should have been set differently.

To your first point, there will be similar changes to this coming so it would be good to standardise on how to report them. I like the idea you propose, but I couldn't find the code you mention. ??

Reply:
Hi David,

> I don't get your second point. First I only abridge when length > max_length. Second adding in the X fudge factor just means max_length should have been set differently.

AFAICS you start abridging if length is exactly max_length. So, you could have:

"Hallo David" maxlen=11 => "Hallo ... omitted 0 characters ... David" 

which is just strange. Additionally, it may seem just strange to replace a small inner portion with a larger "omitted" text, because then the text plus replacement is larger than the original text, e.g. 

"Hallo David" maxlen=10 => "Hallo ... omitted 1 characters ... David" 

So my proposal would be to have a stretch zone the size of the omission text (roughly, modulo variable digits in text) and allow larger text than maxlen but smaller than the stretch zone length. E.g.

"Hallo David" maxlen=10 => "Hallo David"
"Hallo 0123456789012345678901234567890123456789 David" => "Hallo 01234 ... omitted 30 characters ... 56789 David"

Its mainly for optics.

> To your first point, there will be similar changes to this coming so it would be good to standardise on how to report them. I like the idea you propose, but I couldn't find the code you mention. ??

Oh, sorry, this was not in the OpenJDK; this was for a different product I worked on at SAP. But that omission text form served us well.

Reply:
> AFAICS you start abridging if length is exactly max_length

Where do you see that? I have:
```
 bool abridge = length > max_length;
```

>  it may seem just strange to replace a small inner portion with a larger "omitted" text, because then the text plus replacement is larger than the original text,

That is not a practical concern as max_length is expected to be much larger than the replacement text. We don't need the added complexity of the "stretch" zone.

I'm open to changing the omitted text portion though, to include the number of characters omitted.

Reply:
> That is not a practical concern as max_length is expected to be much larger than the replacement text. We don't need the added complexity of the "stretch" zone.

> I'm open to changing the omitted text portion though, to include the number of characters omitted.

Okay, that is fine. I tend to overengineer :)

Thanks for taking my proposal about the omit text form.